presto-main/src/main/java/com/facebook/presto/server/protocol/QueryBlockingRateLimiter.java

```diff
@@ -0,0 +1,106 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server.protocol;
+
+import com.facebook.airlift.stats.CounterStat;
+import com.facebook.presto.execution.QueryManagerConfig;
+import com.facebook.presto.spi.QueryId;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.RateLimiter;
+import com.google.inject.Inject;
+import org.weakref.jmx.Managed;
+import org.weakref.jmx.Nested;
+
+import javax.annotation.PreDestroy;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static java.util.Objects.requireNonNull;
+
+/*
+ * Rate Limiting per query with token bucket
+ * Rate = rateLimitBucketMaxSize/second
+ * When having sufficient tokens, Request will be responded immediately.
+ * When not having enough tokens available, it uses the delayed processing method.
+ */
+public class QueryBlockingRateLimiter
+{
+    private final LoadingCache<QueryId, RateLimiter> rateLimiterCache;
+    private final long rateLimitBucketMaxSize;
+    private final ListeningExecutorService listeningExecutorService;
+
+    private CounterStat rateLimiterTriggeredCounter = new CounterStat();
+
+    @Inject
+    public QueryBlockingRateLimiter(QueryManagerConfig queryManagerConfig)
+    {
+        requireNonNull(queryManagerConfig, "queryManagerConfig is null");
+        this.rateLimitBucketMaxSize = queryManagerConfig.getRateLimitBucketMaxSize();
+        checkArgument(rateLimitBucketMaxSize >= 1, "Rate limit token bucket max size should be at least 1");
+        // Using a custom thread pool with size 1-10 to reduce initial thread resources
+        ExecutorService executorService = new ThreadPoolExecutor(1, 10, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>(), daemonThreadsNamed("rate-limiter-listener"));
+        listeningExecutorService = listeningDecorator(executorService);
```

Comment:
maybe name it as rateLimiterExecutorService

Reply:
will do